### PR TITLE
Big TF test cleanup

### PR DESCRIPTION
**Pull request by Rocketknight1**
Now we've done a big overhaul of the TF model internals, a lot of tests can be fixed. Several tests were disabled for being buggy or too slow - these are almost all performant now, so I re-enabled them. Runtime for the re-enabled tests was 15-20 seconds on my local machine.

Also, we had a number of TF test failures in the daily CI. I think this PR should fix all of them, except for two cases:

Firstly, some models have issues with `resize_token_embeddings`. These failures are caused by the transition to `TFSharedEmbedding` that @gante is currently working on, and I didn't want to interfere! The usual cause is that `resize_token_embeddings` replaces the new-style `TFSharedEmbedding` with an old `tf.Variable`.

Secondly, there are a couple of failures in generate tests. I'm also leaving this to @gante because he knows much more about that code than me :sweat_smile:

### CODE REVIEW

File: src/transformers/models/opt/modeling_tf_opt.py [hunk 1]
```diff
@@ -618,7 +618,7 @@ def call(
             attention_mask = tf.ones(inputs_embeds.shape[:2], dtype=tf.bool)
         else:
             tf.debugging.assert_equal(
-                attention_mask.shape[1],
+                tf.shape(attention_mask)[1],
```
Comment (gante) [inline src/transformers/models/opt/modeling_tf_opt.py:621]: perhaps our good old `shape_list` instead?

File: src/transformers/models/blip/modeling_tf_blip.py [hunk 2]
```diff
@@ -1216,12 +1216,11 @@ def _shift_right(self, input_ids):
     def call(
         self,
         input_ids: tf.Tensor,
-        pixel_values: tf.Tensor,
+        pixel_values: tf.Tensor | None = None,
         decoder_input_ids: tf.Tensor | None = None,
         decoder_attention_mask: tf.Tensor | None = None,
         attention_mask: tf.Tensor | None = None,
         output_attentions: Optional[bool] = None,
-        foutput_attentions: Optional[bool] = None,
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/blip/modeling_tf_blip.py:1224]: ahahaha

**Review by gante — APPROVED** [review]
LGTM 👍

File: src/transformers/models/opt/modeling_tf_opt.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/models/opt/modeling_tf_opt.py:621]: Either would work here! The main thing is that we can't use the static `attention_mask.shape` because it causes compilation to fail as that dimension can be `None`.

**Comment by gante** [comment]
(@Rocketknight1 ping me if the gen tests are not sorted after the latest push)

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -1157,6 +1157,8 @@ def build(self, input_shape=None):
             self.built = True
         else:
             self.built = True
+            # Set the serving spec quickly to ensure that Keras doesn't use the specific dummy input shapes as the spec
+            self._set_save_spec(self._prune_signature(self.input_signature))
```
Comment (ydshieh) [inline src/transformers/modeling_tf_utils.py:1161]: Maybe explain a bit why we don't do this in init method?

File: src/transformers/models/xglm/modeling_tf_xglm.py [hunk 4]
```diff
@@ -463,19 +463,12 @@ def _prepare_decoder_attention_mask(
     ) -> tf.Tensor:
         # create causal mask
         # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
-        combined_attention_mask: tf.Tensor | None = None
-        if input_shape[-1] > 1:
```
Comment (ydshieh) [inline src/transformers/models/xglm/modeling_tf_xglm.py:467]: we don't need to check this condition anymore ..?

File: tests/models/bart/test_modeling_tf_bart.py [hunk 5]
```diff
@@ -225,10 +225,6 @@ def test_decoder_model_past_large_inputs(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs_for_common()
         self.model_tester.check_decoder_model_past_large_inputs(*config_and_inputs)
 
-    @tooslow
-    def test_saved_model_creation(self):
-        pass
-
```
Comment (ydshieh) [inline tests/models/bart/test_modeling_tf_bart.py:231]: This is still a slow test, and I would like to know if this re-enabled test pass now.

File: tests/models/groupvit/test_modeling_tf_groupvit.py [hunk 6]
```diff
@@ -692,11 +692,6 @@ def test_model_from_pretrained(self):
     def test_saved_model_creation(self):
         pass
 
-    @unittest.skip(reason="Currently `saved_model` doesn't work with nested outputs.")
-    @slow
-    def test_saved_model_creation_extended(self):
-        pass
```
Comment (ydshieh) [inline tests/models/groupvit/test_modeling_tf_groupvit.py:698]: Would like to know if this re-enabled test pass now.

File: tests/models/led/test_modeling_tf_led.py [hunk 7]
```diff
@@ -292,11 +292,7 @@ def check_encoder_attentions_output(outputs):
             self.assertEqual(model.config.output_hidden_states, True)
             check_encoder_attentions_output(outputs)
 
-    def test_xla_mode(self):
-        # TODO JP: Make LED XLA compliant
-        pass
```
Comment (ydshieh) [inline tests/models/led/test_modeling_tf_led.py:297]: Would like to know if this re-enabled test pass now.

**Review by ydshieh — APPROVED** [review]
Thank you for all this TF things!

Other than making sure (all) the re-enabled tests will pass now (I guess you already checked them.), I have just 2 nit comments.

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py [hunk 8]
```diff
@@ -195,30 +195,19 @@ def _get_embedding(num_embeddings: int, embedding_dim: int, padding_idx: Optiona
...
-        if max_pos > shape_list(self.embeddings)[0]:
-            self.embedding_weights = self._get_embedding(max_pos + self.offset, self.embedding_dim, self.padding_idx)
-            self.embeddings.assign(self.embedding_weights)
-        return tf.reshape(tf.gather(self.embeddings, tf.reshape(position_ids, (-1,)), axis=0), (bsz, seq_len, -1))
+        # Matt: The PyTorch code does a lot of work to cache the embeddings, setting the cached values as a
+        # model attribute in the forward pass. This is extremely forbidden in TF, which wants forward calls to be
+        # idempotent. TF doesn't need that caching anyway, since it can just store constants during compilation,
+        # so we just remove all of that code.
+        embeddings = self._get_embedding(
+            self.padding_idx + 1 + seq_len + self.offset + past_key_values_length, self.embedding_dim, self.padding_idx
```
Comment (amyeroberts) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:208]: This isn't exactly the same as `past_key_values_length` wasn't added before. This seems more correct but can we run some tests on generation to make sure this works as expected?

File: src/transformers/models/transfo_xl/modeling_tf_transfo_xl.py [hunk 9]
```diff
@@ -588,35 +588,19 @@ def call(
...
-                dec_attn_mask += tf.linalg.band_part(tf.ones([qlen, klen], dtype=tf.int32), -1, -mask_shift_len)
+            mask_shift_len = qlen - tf.nn.relu(mask_len)  # Lazy clamping of negatives to zero
+
+            # Use an indicator variable instead of a conditional to keep the compiler happy
+            lower_mask = tf.linalg.band_part(all_ones, -1, 0) - (
+                tf.linalg.band_part(all_ones, mask_shift_len - 1, 0) * tf.cast(mask_shift_len != 0, tf.int32)
+            )
+            dec_attn_mask = upper_mask + lower_mask
+        else:
+            dec_attn_mask = upper_mask
```
Comment (amyeroberts) [inline src/transformers/models/transfo_xl/modeling_tf_transfo_xl.py:603]: This is a lot easier to understand :)

File: src/transformers/models/xglm/modeling_tf_xglm.py [hunk 10]
```diff
@@ -463,19 +463,12 @@ def _prepare_decoder_attention_mask(
...
-            combined_attention_mask = _make_causal_mask(input_shape, past_key_values_length)
-
-        if attention_mask is not None:
+        combined_attention_mask = _make_causal_mask(input_shape, past_key_values_length)
+        if attention_mask is None:
+            return combined_attention_mask
+        else:
```
Comment (amyeroberts) [inline src/transformers/models/xglm/modeling_tf_xglm.py:469]: ultra nit: if we return in an if statement, we don't need the `else`

File: src/transformers/models/xglm/modeling_tf_xglm.py (hunk 4 above)
Comment (amyeroberts) [reply src/transformers/models/xglm/modeling_tf_xglm.py:467]: +1

File: tests/models/lxmert/test_modeling_tf_lxmert.py [hunk 11]
```diff
@@ -581,73 +580,6 @@ def test_compile_tf_model(self):
             extended_model = tf.keras.Model(inputs=[input_ids, visual_feats, visual_pos], outputs=[outputs])
             extended_model.compile(optimizer=optimizer, loss=loss, metrics=[metric])
 
-    @tooslow
-    def test_saved_model_creation(self):
-        pass
-
-    @slow
-    def test_saved_model_creation_extended(self):
```
Comment (amyeroberts) [inline tests/models/lxmert/test_modeling_tf_lxmert.py:589]: Does this run now with the default `test_saved_model_creation_extended` test?

**Review by amyeroberts — COMMENTED** [review]
Thanks for cleaning up! 

Overall, changes look good to me. 
* Big +1 to all of @ydshieh's comments
* For just the affected models, could you run the slow tests this changes? In particular `test_saved_model_creation_extended`?
* Could you run a generation test with speech to text to make sure the embeddings reshaping is working?

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py (hunk 8 above)
Comment (Rocketknight1) [reply src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:208]: Hi @amyeroberts, you're right! It's actually okay to sometimes generate too many embeddings, though, because the embeddings tensor is only transiently created in this function, gathered from and then discarded again. I ran the slow tests for this model and all passed.

Comment (Rocketknight1) [reply src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:208]: ...though as I write this, I realize that all of this code is just a speed hack because eager Torch code can't optimize or compute things out of order, so really I should just directly transform the position IDs into the embeddings and skip the whole gather!

File: src/transformers/modeling_tf_utils.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:1161]: It's a bit of a long story! `_set_save_spec` is normally called internally by Keras, and for subclassed models (i.e. all models in `transformers`) it uses **the first input shapes the model sees.** This was a huge problem for us, because we'd pass in some tiny dummy inputs and it would just lock in that useless specific shape as the model's save spec. This made exporting/serving a real nightmare!

We avoid that by setting a correct, general save spec before the model has seen any inputs. It doesn't really matter in most cases whether we put that in the `__init__` or the `build` method, as long as it happens before we pass dummy inputs in. However, there is one edge case where it makes a small difference: If the user builds a model from a config, and then passes inputs of a specific shape in. In this case, putting it in `build()` allows the user to set the save spec with their own inputs, which can be useful in a couple of cases.

I'm not convinced this is a perfect solution, but it resolves an edge case in our in-graph tokenizer test, so it seems a little better than the alternative!

File: tests/models/bart/test_modeling_tf_bart.py (hunk 5 above)
Comment (Rocketknight1) [reply tests/models/bart/test_modeling_tf_bart.py:231]: It passes!

File: src/transformers/models/xglm/modeling_tf_xglm.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/models/xglm/modeling_tf_xglm.py:467]: I don't think so! I couldn't see a case where `input_shape[-1] == 0` was possible.

File: tests/models/rag/test_modeling_tf_rag.py [hunk 12]
```diff
@@ -490,6 +490,7 @@ def test_model_without_retriever(self):
         inputs_dict = self.config_and_inputs
         self.check_model_without_retriever(**inputs_dict)
 
+    @slow
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/models/rag/test_modeling_tf_rag.py:493]: These aren't failures! They're just extremely slow generation tests (with retrieval!), and were sometimes triggering the 120s timeout in the live CI.

File: tests/models/lxmert/test_modeling_tf_lxmert.py (hunk 11 above)
Comment (Rocketknight1) [reply tests/models/lxmert/test_modeling_tf_lxmert.py:589]: `test_saved_model_creation_extended` is now a core test that is only run on a few models because it's very expensive, so this skip is no longer needed.

File: src/transformers/models/xglm/modeling_tf_xglm.py (hunk 10 above)
Comment (Rocketknight1) [reply src/transformers/models/xglm/modeling_tf_xglm.py:469]: Fixed!

File: src/transformers/models/xglm/modeling_tf_xglm.py (hunk 4 above)
Comment (ydshieh) [reply src/transformers/models/xglm/modeling_tf_xglm.py:467]: I mean `> 1` vs `== 1` not `== 0`

File: src/transformers/modeling_tf_utils.py (hunk 3 above)
Comment (ydshieh) [reply src/transformers/modeling_tf_utils.py:1161]: So a short version could be 
```python
# put this in build() allows the user to set the save spec with their own inputs.
```
😄 ...?

File: src/transformers/models/xglm/modeling_tf_xglm.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/models/xglm/modeling_tf_xglm.py:467]: Ah, you're totally right, I don't know how I blanked on that! Let me fix it.

File: src/transformers/modeling_tf_utils.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:1161]: Sure! I'll comment something like that

File: tests/models/mobilebert/test_modeling_tf_mobilebert.py [hunk 13]
```diff
@@ -311,15 +311,6 @@ def test_for_token_classification(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs()
         self.model_tester.create_and_check_mobilebert_for_token_classification(*config_and_inputs)
 
-    @slow
-    def test_keras_fit(self):
```
Comment (amyeroberts) [inline tests/models/mobilebert/test_modeling_tf_mobilebert.py:315]: Just double checking that this is now fast for this model?

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating! 

I just have one question to address re the refactoring of `_prepare_decoder_attention_mask` before merging

File: tests/models/mobilebert/test_modeling_tf_mobilebert.py (hunk 13 above)
Comment (Rocketknight1) [reply tests/models/mobilebert/test_modeling_tf_mobilebert.py:315]: Actually, good question. On my local machine, this test takes from 20-40 seconds depending on model. MobileBERT is one of the slower ones, but it's still inside that range.

However, 20-40 seconds is probably in the range that the whole test should be marked as slow to keep it out of the quick CI, right?

File: src/transformers/models/xglm/modeling_tf_xglm.py [hunk 14]
```diff
@@ -463,19 +463,14 @@ def _prepare_decoder_attention_mask(
...
-            combined_attention_mask = (
-                expand_attention_mask
-                if combined_attention_mask is None
-                else expand_attention_mask + combined_attention_mask
-            )
-
-        return combined_attention_mask
+        combined_attention_mask = _make_causal_mask(input_shape, past_key_values_length)
+        combined_attention_mask = tf.cond(
+            input_shape[-1] > 1, lambda: combined_attention_mask, lambda: tf.ones_like(combined_attention_mask)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/xglm/modeling_tf_xglm.py:468]: Ah, I think you're right! Hang on, let me see what I can do.

File: tests/models/mobilebert/test_modeling_tf_mobilebert.py (hunk 13 above)
Comment (ydshieh) [reply tests/models/mobilebert/test_modeling_tf_mobilebert.py:315]: We have some tests (not decorated as slow) run more than 40 seconds, but let's not add more such tests. Decorate it as slow and everyone's life is easier 🍺

Comment (amyeroberts) [reply tests/models/mobilebert/test_modeling_tf_mobilebert.py:315]: 🍻  cheers to that!

File: src/transformers/models/xglm/modeling_tf_xglm.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/xglm/modeling_tf_xglm.py:468]: Investigation complete! So basically, I can't actually make this function reproduce the old behaviour when we're compiling with flexible shapes, because it's just totally forbidden in TF to have a conditional where one branch returns `None` and the other branch returns a `tf.Tensor`. 

However, when I actually followed the code through to self-attention where the attention mask is used, an attention mask of `None` is just treated as an all-ones mask (i.e. neither affects the attention logits at all). Therefore, returning all-ones instead of `None` yields the same model outputs, while obeying TF's requirements for compiling conditionals.

**Comment by Rocketknight1** [comment]
I think everything has been addressed now, but I'm not going to merge this one today because there's another PR affecting our tests (#24301) and ideally I'd like to be able to separately view their impact on the CI!

**Comment by ydshieh** [comment]
> I think everything has been addressed now, but I'm not going to merge this one today

Nice 👍 .

I never merge PRs on Firday evening or early afternoon. I don't want to get a ☎️ ⚡ !

**Comment by ydshieh** [comment]
Wait, you merged ...!? (but you said you are not going to merge 🤔 )
